### PR TITLE
Add solution bounds to the box constraints

### LINKED ISSUE #70: Generic Box Constraint
Right now, the box constraints with $l_i = 0$ and $u_i = 1$ is hard coded. We should make this to be general with $l_i$ and $u_i$ defined at the time of defining the `BoxQP` problem instances with default values of $0$ and $1$.  

This change includes updating all the functions that calculate the gradient of the objective function.

Currently the code in the `_change_variable_boxqp` method looks like this:
`return 0.5 * (problem_variables / S + 1)`

It should look something like this:
`return 0.5 * problems_variables / S * (upper_limit - lower_limit) + 0.5 * (upper_limit + lower_limit)`. 

The functions `_calculate_grads_boxqp` and `_calculate_drift_boxqp` also needs to be updated to something like this:
`c_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", c * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit)/ S`
`c_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)`

### PR DESCRIPTION
**Pull request by KatieO4t1qbit**
### Description

The upper and lower limit of the box constraints were being implicitly factored into the equations as though the lower limit was 0 and the upper limit was 1. To make the code more versatile, we've made the equations explicit and added the solution_bounds as a parameter to the ProblemInstance class.

### References

Closes #70 

### Testing

<TBD>

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `main`

### CODE REVIEW

File: ccvm_simulators/solvers/dl_solver.py [hunk 1]
```diff
@@ -114,7 +114,9 @@ def parameter_key(self, parameters):
         self._parameter_key = parameters
         self._is_tuned = False
 
-    def _calculate_drift_boxqp(self, c, s, pump, rate, feedback_scale=100, S=1):
+    def _calculate_drift_boxqp(
```
Comment (KatieO4t1qbit) [inline ccvm_simulators/solvers/dl_solver.py:117]: @FarhadK-1QBit 
This is from the issue description:
```
The functions _calculate_grads_boxqp and _calculate_drift_boxqp also needs to be updated to something like this:
c_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", c * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit)/ S
c_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
```

But I am not confident how to apply this to `_calculate_grads_boxqp`, can you please help me with that?

Also, do the s_grad variables need to be updated to account for upper and lower bounds, too?

Comment (KatieO4t1qbit) [reply ccvm_simulators/solvers/dl_solver.py:117]: Similarly, can we discuss the implementation for the other solvers' functions?

Comment (FarhadK-1QBit) [reply ccvm_simulators/solvers/dl_solver.py:117]: For `_calculate_grads_boxqp` exactly the same happens, that is, in this line: https://github.com/1QB-Information-Technologies/ccvm/blob/b0b2c29ad28f1dae4fb385b89f9468c349548dac/ccvm_simulators/solvers/dl_solver.py#L180

Replace the code with this:
```
c_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", c * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit)/ S
c_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
```

For both `_calculate_drift_boxqp` and `_calculate_grads_boxqp` we can write similarly for the `s_grad`:
```
s_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", s * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit)/ S
s_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
```

In the MF-CCVM, you can use this for both `_calculate_drift_boxqp` and `_calculate_grads_boxqp` here: 
https://github.com/1QB-Information-Technologies/ccvm/blob/b0b2c29ad28f1dae4fb385b89f9468c349548dac/ccvm_simulators/solvers/mf_solver.py#L201C1-L207C1

```
mu_term2_1 = ( -(1 / 4) * (torch.einsum("bi,ij -> bj", mu_tilde * (upper_limit - lower_limit) / S  + (upper_limit + lower_limit), self.q_matrix)) * (upper_limit - lower_limit) / S )
mu_term2_2 = -self.v_vector * (upper_limit - lower_limit) / (2 * S)
```

For the Langevin solver you can write again for both `_calculate_drift_boxqp` and `_calculate_grads_boxqp`:
```
c_drift_1 = torch.einsum("bi,ij -> bj", c *(upper_limit - lower_limit)/ (2 * S) + (upper_limit + lower_limit)/ 2, self.q_matrix)
c_drift_2 = self.v_vector

c_drift = -(c_drift_1 + c_drift_2) * (upper_limit - lower_limit) / (2 * S)
```

And for pumped-Langevin you can write for `_calculate_grads_boxqp` here: 
https://github.com/1QB-Information-Technologies/ccvm/blob/b0b2c29ad28f1dae4fb385b89f9468c349548dac/ccvm_simulators/solvers/pumped_langevin_solver.py#L129C1-L132C43

```
c_grad_1 = torch.einsum("bi,ij -> bj", c * (upper_limit - lower_limit) / (2 * S) + (upper_limit + lower_limit) / 2, self.q_matrix) * (upper_limit - lower_limit) / (2 * S)
c_grad_2 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
```
The `_calculate_drift_boxqp` doesn't need to be changed here. 

Sorry for not clarifying this before. Let me know if you're unsure about any parts. I will make sure to review these updates.

Thanks so much!

**Review by FarhadK-1QBit — COMMENTED** [review]
Thanks for the PR, @KatieO4t1qbit ! I answered your question. Let me know when you've implemented them.

File: ccvm_simulators/solvers/dl_solver.py (hunk 1 above)
Comment (KatieO4t1qbit) [reply ccvm_simulators/solvers/dl_solver.py:117]: Thank you for providing all of these implementations! I appreciate that!

**Review by Estella-Wang-1qbit — APPROVED** [review]
I have no further comments regarding the code. The changes and added tests look good to me. 👍 

One thing for your consideration is that it would be nice to add a sentence to the CHANGELOG informing users of this change, so they know they can set the bounds when initializing the problem instance.

File: ccvm_simulators/solvers/dl_solver.py [hunk 2]
```diff
@@ -158,14 +180,25 @@ def _calculate_grads_boxqp(self, c, s, S=1):
...
+                c * (upper_limit - lower_limit) / S + (upper_limit + lower_limit),
+                self.q_matrix,
+            )
+            * (upper_limit - lower_limit)
+            / S
+        )
+        c_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
 
         s_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", s / S + 1, self.q_matrix) / S
         s_grad_3 = self.v_vector / 2 / S
```
Comment (FarhadK-1QBit) [inline ccvm_simulators/solvers/dl_solver.py:204]: The `s_grad` variables should also be updated similarly here:
```suggestion
        s_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", s * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit) / S
        s_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2*S)
```

**Review by FarhadK-1QBit — CHANGES_REQUESTED** [review]
This looks great, Katie! There is just one little spot that has been missed. After that, it's ready to be merged. 

Thank you!

File: ccvm_simulators/solvers/dl_solver.py (hunk 2 above)
Comment (KatieO4t1qbit) [reply ccvm_simulators/solvers/dl_solver.py:204]: Ahh, I missed them! Thanks for noticing.

Comment (KatieO4t1qbit) [reply ccvm_simulators/solvers/dl_solver.py:204]: What about the calculate_grads function? It also has these lines but I haven't updated them

**Review by FarhadK-1QBit — APPROVED** [review]
Looks great, Katie! Thanks so much!!

**Pull request merged**